### ISSUE TITLE
Determine how many bits in pointers we can safely use for tags

### ISSUE THREAD

**Issue by punchfox** [spotify/spotify-json#43]
The only way we can get reasonable performance and memory usage from a graph based API is to use tagged pointers and nanboxing to store the values in 64-bit words. It's important that we follow the linear address space rules for each platform, so that we don't end up with invalid pointers after an operating system or update. The spotify-json library is deployed on literary hundreds of millions of devices, from desktop systems to mobile phones to servers, some harder to upgrade than others. We need to ensure that we are correct on our four supported platforms: ARM, ARM-64, x86, x86-64.

**Comment by punchfox** [comment]
- [x] ARM-32
- [ ] ARM-64
- [x] x86-32
- [ ] x86-64

**Comment by punchfox** [comment]
The 32-bit platforms are obviously ok. We have more than enough bits to store the whole range (32 bits) in a nanboxed 64-bit word (at least 48 bits).

**Comment by punchfox** [comment]
The 64 bit ARM architecture specifies that the top 16 bits must either be `0x0000` or `0xFFFF`, with sign extension of the 48-bit address. This works well for purposes of nanboxing. There's a feature that allows the CPU to ignore the top 8-bits, but we don't need to use that.

https://developer.arm.com/docs/den0024/latest/12-the-memory-management-unit/125-translation-table-configuration/1251-virtual-address-tagging

**Comment by punchfox** [comment]
The situation on x86-64 is a bit more tricky. Up until recently, the 4-level paging architecture on Intel CPUs allowed for 48-bit linear addresses, with the same sign extension scheme as used on 64-bit ARM. Intel is adding 5-level paging though, which increases the linear address space to 57 bits. This may cause a problem for us. If we align our values to 8 bytes (64 bits), we can shift those 57 bits into 54 bits, which is still a lot of bits to store.

https://software.intel.com/sites/default/files/managed/2b/80/5-level_paging_white_paper.pdf

**Comment by punchfox** [comment]
For reference, this is the layout of a 64-bit NaN (in IEEE 754):

`x1111111 1111xxxx xxxxxxxx xxxxxxxx`

The 11 exponent bits are all set to 1s, leaving a total of 53 free bits, where 52 of those bits are continuous. We have four value types that will be heap allocated: strings, arrays and objects (maps), and 64-bit integers. We can either use two of those 53 bits for the type, or store the type in the heap allocated structure. The former would be better for performance, while the latter might be what we actually need to do, considering that we need to stuff 57 bits addresses into 53 bits.

**Comment by punchfox** [comment]
The allocation quantum size on various operating systems:

32-bit macOS: 16 bytes
64-bit macOS: 32 bytes

Source: [Mac OS X Internals: A Systems Approach](https://books.google.se/books?id=K8vUkpOXhN4C&pg=PA956&lpg=PA956&dq=mac+os+malloc+quantum+size&source=bl&ots=OLqhR-XtXv&sig=CiZ98eGwfqQdEoILhFQBFHKGMwg&hl=en&sa=X&ved=0ahUKEwja1q21xafTAhWidpoKHdwVAx8Q6AEIKDAB#v=onepage&q=mac%20os%20malloc%20quantum%20size&f=false)

The quantum size does not necessarily imply alignment though, but it should be fairly safe to assume that most operating systems align the addresses to 16 bytes, so that they can be used with 128-bit vectors, e.g., SSE and NEON.

By allocating our heap structures with 16 byte alignment, we can shift away the bottom four bits on our 57-bit addresses, giving us 53-bits of address data, which is exactly how many free bits we have in our nanboxed 64-bit words.

**Comment by punchfox** [comment]
We might as well ask for 32-byte aligned memory, so that we can use the 52 continuous bits and leave the sign bit for something else, perhaps to indicate _shortstrings_ stuffed in those 52 bytes. As we discussed IRL, the addresses into the _null page_ can be used for special values, such as `null`, `false`, and `true`. For integers, regular doubles can store integers less than 2^53, and then we'll have to switch to a heap allocated `int64_t` for anything larger.

**Comment by fxb** [comment]
It seems ```alignof(std::max_align_t)``` is ```16``` on macOS 64-bit. With C11 / C++17 we could use ```aligned_alloc```...

**Comment by punchfox** [comment]
If we can't find any way to allocate aligned to 32 bytes on all platforms, we could always ask for 16 bytes more than we need, offset the pointer with 16 bytes if we got a non-aligned address. Since we need to have a header for the value type anyway (no bits left in the nanboxed pointer), we can use one bit there to indicate if the real address is 16 bytes back, for deallocation purposes. On macOS 64-bit it doesn't matter anyway if we allocate 16 bytes or 32 bytes; we still end up with a 32 byte allocation.

```c++
enum class value_type {
  array, object, string, integer
};

struct value_header_64 {
  bool offset_by_16: 1;
  value_type type: 2;
  size_t size: 61;  // not used for integers
};
```

Since the size of the `value_header_64` is only 8 bytes, we can follow up with another 8 bytes for the  `int64_t` of the `integer` type, totalling 16 bytes, staying within our 32 byte allocation.

```
   aligned: [ vheader ] [ int64_t ] [ ------- ] [ ------- ]
misaligned: [ ------- ] [ ------- ] [ vheader ] [ int64_t ]
```

**Comment by punchfox** [comment]
[This](https://lwn.net/Articles/718888/) ARM security feature makes things yet more complicated. It uses the top unused bits to sign the pointer, to reduce the risk of buffer overflow attacks. Currently, compilers (GCC) only use this for return addresses, but that could certainly change.

We discussed this offline and we have decided to actually just go with 16 byte values. The code will be much simpler, there is no risk of problems in the future, and since JSON is so string heavy, we will probably benefit from having 15 character shortstrings instead of only 5 or so.
